### PR TITLE
8308479: [s390x] Implement alternative fast-locking scheme

### PR DESCRIPTION
**Pull request by offamitkumar**
This PR implements new fast-locking scheme for s390x. Additionally few parameters have been renamed to be in sync with PPC.

Testing done (for release, fastdebug and slowdebug build):
All `test/jdk/java/util/concurrent` test with parameters:
* LockingMode=2 
* LockingMode=2 with -Xint
* LockingMode=2 with -XX:TieredStopAtLevel=1
* LockingMode=2 with -XX:-TieredCompilation

Result is consistently similar to Aarch(MacOS) and PPC, All of 124 tests are passing except `MapLoops.java` because in the 2nd part for this testcase, jvm starts with `HeavyMonitors` which conflict with `LockingMode=2`

BenchMark Result for Renaissance-jmh:

| Benchmark                                | Without fastLock (ms/op) | With fastLock (ms/op) | Improvement |
|------------------------------------------|-------------------------|----------------------|-------------|
| o.r.actors.JmhAkkaUct.runOperation        | 1565.080                | 1365.877             | 12.70%      |
| o.r.actors.JmhReactors.runOperation       | 9316.972                | 10592.982            | -13.70%     |
| o.r.jdk.concurrent.JmhFjKmeans.runOperation | 1257.183                | 1235.530             | 1.73%       |
| o.r.jdk.concurrent.JmhFutureGenetic.runOperation | 1925.158                | 2073.066             | -7.69%      |
| o.r.jdk.streams.JmhParMnemonics.runOperation | 2746.664                | 2836.085             | -3.24%      |
| o.r.jdk.streams.JmhScrabble.runOperation   | 76.774                  | 74.239               | 3.31%       |
| o.r.rx.JmhRxScrabble.runOperation          | 162.270                 | 167.061              | -2.96%      |
| o.r.scala.sat.JmhScalaDoku.runOperation    | 3333.711                | 3271.078             | 1.88%       |
| o.r.scala.stdlib.JmhScalaKmeans.runOperation | 182.746                 | 182.153              | 0.33%       |
| o.r.scala.stm.JmhPhilosophers.runOperation | 15003.329               | 13396.921            | 10.57%      |
| o.r.scala.stm.JmhScalaStmBench7.runOperation | 1669.090                | 1579.900             | 5.34%       |
| o.r.twitter.finagle.JmhFinagleChirper.runOperation | 9601.963                | 10034.404            | -4.52%      |
| o.r.twitter.finagle.JmhFinagleHttp.runOperation | 4403.725                | 4746.707             | -7.79%      |

DaCapo Benchmark Result:

| Benchmark                | Without fast lock (msec) | With fast lock (msec) | Improvement |
|--------------------------|-------------------------|----------------------|-------------|
| DaCapo 9.12 h2 Run 1     | 117,010                 | 108,699              | 7.10%       |
| DaCapo 9.12 h2 Run 2     | 111,986                 | 107,373              | 4.11%       |

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308479](https://bugs.openjdk.org/browse/JDK-8308479): [s390x] Implement alternative fast-locking scheme (**Enhancement** - P4)

### Reviewers
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14414/head:pull/14414` \
`$ git checkout pull/14414`

Update a local copy of the PR: \
`$ git checkout pull/14414` \
`$ git pull https://git.openjdk.org/jdk.git pull/14414/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14414`

View PR using the GUI difftool: \
`$ git pr show -t 14414`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14414.diff">https://git.openjdk.org/jdk/pull/14414.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14414#issuecomment-1587115712)

### CODE REVIEW

**Comment by offamitkumar** [comment]
Hi @RealLucy, @TheRealMDoerr, @reinrich 
Please review this PR as per your availability. 
Thank you.

**Review by RealLucy — CHANGES_REQUESTED** [review]
There are some comments you may want to consider.

File: src/hotspot/cpu/s390/c1_MacroAssembler_s390.hpp [hunk 1]
```diff
@@ -51,6 +51,7 @@
   // hdr     : Used to hold original markWord to be CASed back into obj, contents destroyed.
   // obj     : Must point to the object to lock, contents preserved.
   // disp_hdr: Must point to the displaced header location, contents destroyed.
+  // Z_R0_scratch will be killed as well
```
[the remark this answers is not in the document]
Comment (offamitkumar) [reply src/hotspot/cpu/s390/c1_MacroAssembler_s390.hpp]: Thanks for all suggestions (and expecting more), I've updated the code.

Please consider looking at the latest changes. :-)

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 2]
```diff
@@ -5621,3 +5642,107 @@ SkipIfEqual::SkipIfEqual(MacroAssembler* masm, const bool* flag_addr, bool value
...
+  compareU32_and_branch(temp, (unsigned)LockStack::end_offset()-1, bcondHigh, slow_case);
+
+  // attempting a fast_lock
+  // Load (object->mark() | 1) into hdr
+  z_oill(hdr, markWord::unlocked_value);
+
+  z_lgr(temp, hdr);
+
+  // Clear lock-bits from hdr (locked state)
+  z_xilf(temp, markWord::unlocked_value);
```
[the remark this answers is not in the document]
Comment (offamitkumar) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:5670]: we don't have `z_xill()` instruction like `z_nill()` or `z_oill()` :-(

[the remark this answers is not in the document]
Comment (RealLucy) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:5670]: You are right. I should have RTFM.

**Comment by offamitkumar** [comment]
Hi @RealLucy,

Please check benchmark result in description.

**Comment by TheRealMDoerr** [comment]
Are the benchmark results stable or do they have a large variance?
Did you also compare against the version without your changes? That should be done for sanity checking. You have modified the code for the legacy mode, too.
I'm not sure what you mean by "With/Without fastLock"? Patch applied? LockingMode selection?

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 3]
```diff
@@ -3251,27 +3259,40 @@ void MacroAssembler::compiler_fast_unlock_object(Register oop, Register box, Reg
 
   // Handle existing monitor.
   // The object has an existing monitor iff (mark & monitor_value) != 0.
-  z_lg(currentHeader, oopDesc::mark_offset_in_bytes(), oop);
+  z_lg(currentHeader, hdr_offset, oop);
   guarantee(Immediate::is_uimm16(markWord::monitor_value), "must be half-word");
+  if (LockingMode == LM_LIGHTWEIGHT)
+    z_lgr(temp, currentHeader);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/macroAssembler_s390.cpp:3265]: Hotspot coding style: Use curly brackets even if it's only one statement.

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 4]
```diff
@@ -3251,27 +3259,40 @@ void MacroAssembler::compiler_fast_unlock_object(Register oop, Register box, Reg
...
-    z_bru(done); // Csg sets CR as desired.
+    z_bru(done); // csg sets CR as desired.
   } else {
-    // Set NE to indicate 'failure' -> take slow-path
-    z_ltgr(oop, oop);
+    assert(LockingMode == LM_LIGHTWEIGHT, "must be");
+
+    // don't load currentHead again from stack-top after monitor check, as it is possible
+    // some other thread modified it.
+    z_lgr(currentHeader, temp);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/macroAssembler_s390.cpp:3285]: Moving forth and back is not needed. E.g. register usage can be swapped.

**Review by TheRealMDoerr — COMMENTED** [review]
Minor suggestions.

**Comment by offamitkumar** [comment]
>Are the benchmark results stable or do they have a large variance?

I ran `DaCapo` twice, but with `Renaissance-jmh` it was just a single run. 

>Did you also compare against the version without your changes? That should be done for sanity checking. You have modified the code for the legacy mode, too.
>I'm not sure what you mean by "With/Without fastLock"? Patch applied? LockingMode selection?

Patch was applied, but for in first case LockingMode was not selected, and in second case (with fast locking) `LockingMode=2` argument was given. 

But I will run these again with/without the code change. Thanks for insights.

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 3 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:3265]: Thanks for comments, 

Fixed.

**Comment by TheRealMDoerr** [comment]
> After struggling a lot with my environment, I finally was able to run some performance tests. There is the "db" subtest of JVM98 which caught my attention. It reproducibly shows
> 
> * LockingMode=0: avg time: 2888ms
> * LockingMode=1: avg time: 867ms
> * LockingMode=2: avg time: 1020ms
> 
> We need to gain deeper insight into the why before integration.

Note that the new fast locking (LockingMode 2) doesn't support fast recursive locking. However, I doubt that this is the problem, here. LockingMode 2 is fastest on Power10 for "db" benchmark. So, yes, interesting finding!

**Comment by offamitkumar** [comment]
with 

fastlockbench:

|                 | Time (ns/op) [Average] | Performance Improvement |
|-----------------|-----------------------|-------------------------|
| With Patch      | 18.614                | -                       |
| Without Patch (+LockingMode=2)   | 19.116                | -                       |
| Improvement     | -                     | 2.63%                   |


Dacapo:

|                 | Time (msec)            | Performance Improvement |
|-----------------|-----------------------|-------------------------|
| Without Patch   | 118199                | -                       |
| With Patch (+LockingMode=2)      | 120213                | -                       |
| Improvement     | -                     | -1.70%                  |



 Renaissance-jmh:
| Benchmark                                           | Without Patch | With Patch (+LockingMode=2) | Improvement (%) |
|-----------------------------------------------------|--------------|-------------|-----------------|
| o.r.actors.JmhAkkaUct.runOperation                  | 1349.563     | 1456.927    | -8.00           |
| o.r.actors.JmhReactors.runOperation                 | 10296.818    | 11605.138   | -12.82          |
| o.r.jdk.concurrent.JmhFjKmeans.runOperation         | 1145.668     | 1390.726    | -21.35          |
| o.r.jdk.concurrent.JmhFutureGenetic.runOperation    | 1952.648     | 1976.534    | -1.22           |
| o.r.jdk.streams.JmhParMnemonics.runOperation        | 2819.193     | 2801.261    | 0.64            |
| o.r.jdk.streams.JmhScrabble.runOperation            | 78.594       | 74.529      | 5.16            |
| o.r.rx.JmhRxScrabble.runOperation                   | 171.220      | 180.978     | -5.70           |
| o.r.scala.sat.JmhScalaDoku.runOperation             | 5841.122     | 3556.529    | 39.15           |
| o.r.scala.stdlib.JmhScalaKmeans.runOperation        | 199.683      | 183.357     | 8.18            |
| o.r.scala.stm.JmhPhilosophers.runOperation          | 14482.716    | 15834.972   | -9.32           |
| o.r.scala.stm.JmhScalaStmBench7.runOperation        | 1567.814     | 1716.439    | -9.46           |
| o.r.twitter.finagle.JmhFinagleChirper.runOperation  | 9477.834     | 9737.119    | -2.73           |
| o.r.twitter.finagle.JmhFinagleHttp.runOperation     | 4381.338     | 4681.571    | -6.85           |


These are the new results I'm getting, seems consistent as well. `-21.35 ` appears to be a lot but not sure if it's of observable category or not.

**Review by TheRealMDoerr — APPROVED** [review]
I think my concerns had been addressed. The performance topic can be investigated later if there's not enough time right now. I think it's good enough for an experimental feature. What's your opinion, @RealLucy?

**Comment by coleenp** [comment]
We are considering making Fast Locking on by default for Oracle supported platforms.  Have these performance concerns been addressed?

**Comment by RealLucy** [comment]
> We are considering making Fast Locking on by default for Oracle supported platforms. Have these performance concerns been addressed?

The performance regression which is observed in some tests is still not fully understood. I will approve the PR despite of that. According to all out testing, it is functionally correct.

**Comment by offamitkumar** [comment]
Lutz, Martin, Thanks for reviews and help. 

/integrate

**Pull request closed without merging**